File: vokimi_frontend/src/ts/page_classes/manage_test_page/conclusion/FeedbackRecordData.ts
import { StringUtils } from "../../../utils/StringUtils";

export class FeedbackRecordData {
    recordId: string;
    text: string;
    createdAt: Date;
    userId: string | null;
    authorUsername: string;
    authorProfilePicture: string;

    constructor(
        recordId: string,
        text: string,
        date: Date | string,
        authorId: string | null,
        authorUsername: string,
        authorProfilePicture: string
    ) {
        this.recordId = recordId;
        this.text = text;
        this.createdAt = date instanceof Date ? date : new Date(date);
        this.userId = authorId;
        this.authorUsername = authorUsername;
        this.authorProfilePicture = authorProfilePicture;
    }
    isAnonymous(): boolean {
        return StringUtils.isNullOrWhiteSpace(this.userId)
    }
    static fromResponseData(data: any): FeedbackRecordData {
        return new FeedbackRecordData(
            data.recordId,
            data.text,
            data.createdAt,
            data.userId,
            data.authorUsername,
            data.authorProfilePicture
        );
    }
}

File: vokimi_api/Src/dtos/responses/manage_test_page/conclusion/FeedbackRecordData.cs
using vokimi_api.Src.constants_store_classes;
using vokimi_api.Src.db_related.db_entities.tests_related;

namespace vokimi_api.Src.dtos.responses.manage_test_page.conclusion
{
    public record class FeedbackRecordData(
        string RecordId,
        string Text,
        DateTime CreatedAt,
        string? UserId,
        string AuthorUsername,
        string AuthorProfilePicture
    )
    {
        public static FeedbackRecordData FromTestFeedbackRecord(TestFeedbackRecord feedbackRecord) {
            if (feedbackRecord.IsAnonymous) {
                return Anonymous(feedbackRecord);
            }
            if (feedbackRecord.UserId is not null && feedbackRecord.AppUser is not null) {
                return WihtUser(feedbackRecord);
            }
            throw new InvalidOperationException("Invalid feedback record state.");
        }
        private static FeedbackRecordData Anonymous(TestFeedbackRecord f) =>
            new(f.Id.ToString(), f.Text, f.CreatedAt, null, "Anonymous user", ImgOperationsConsts.AnonymousProfilePicture);
        private static FeedbackRecordData WihtUser(TestFeedbackRecord f) =>
            new(f.Id.ToString(), f.Text, f.CreatedAt, f.AppUser.Id.ToString(), f.AppUser.Username, f.AppUser.ProfilePicturePath);
    }

}
